### PR TITLE
 battery-util: debug log when we're unsure about the battery level 

### PR DESCRIPTION
**Pull request by YHNdnzj**

### CODE REVIEW

**Comment by medhefgo** [comment]
Would summing the total charge level be a good idea?

**Review by bluca — APPROVED** [review]

**Comment by keszybz** [comment]
> I don't think we should blindly say that we're not in low battery state just because there's one battery whose state can't be read. Instead, let's continue to check other batteries, and reach that conclusion only if all of them are also unreadable or some of them are charged.

This description doesn't actually match *status quo ante*. The code would actually continue if there's a read failure, and return true only we found at least one batter in low state. Can you show a specific example where the old code would behave worse than the patched code?

**Comment by YHNdnzj** [comment]
> The code would actually continue if there's a read failure, and return true only we found at least one batter in low state.

Well, it does say `continue` after setting `unsure` to `true`, but we then return `false` straightly as the loop finishes if `unsure` is `true`. Effectively we simply return `false` if there's any existing battery whose state can't be read, which I consider spurious here.

**Comment by YHNdnzj** [comment]
> Would summing the total charge level be a good idea?

Hmm, I'm not sure if something like that is actually needed? If there're multiple batteries but none of them has a capacity higher than the threshold of 5%, I'd say we can consider the system in battery low state too.

**Comment by keszybz** [comment]
> Well, it does say `continue` after setting `unsure` to `true`, but we then return `false` straightly as the loop finishes if `unsure` is `true`.

Where do you see this?
https://github.com/systemd/systemd/blob/c41e3605ec2f5c1d3040aa3a5fd57b9aee08c198/src/shared/battery-util.c#L254-L267

**Comment by YHNdnzj** [comment]
> Where do you see this?

We return `false` if `unsure` is true, regardless of `found_low`:

https://github.com/systemd/systemd/blob/c41e3605ec2f5c1d3040aa3a5fd57b9aee08c198/src/shared/battery-util.c#L269-L271

**Comment by poettering** [comment]
So fundamentally this just boils down to what to assume if we know there's a battery but we cannot read it. Should we assume then that we are in a low power state or in a regular state. Since the low power state is conceptually supposed to be the exception I'd always say we should default to "not low power state" in that case, which is what is implemented right now, no?

**Comment by poettering** [comment]
i cherrypicked the 2nd commit into git main already, since it should be uncontroversial

**Comment by YHNdnzj** [comment]
> which kinda makes sense to me: don't assume the worst if we have no clear indication.

OK. I originally thought that we shouldn't be that positive...

I've reworked this to generate a debug log when we're unsure about the battery level. PTAL.

File: src/shared/battery-util.c [hunk 1]
```diff
@@ -267,13 +267,11 @@ int battery_is_discharging_and_low(void) {
         }
 
         /* If we found a battery whose state we couldn't read, don't assume we are in low battery state */
-        if (unsure)
+        if (unsure) {
+                log_debug("Found battery with unreadable state, assuming not in low battery state.");
```
Comment (poettering) [inline src/shared/battery-util.c:271]: log_device_debug(dev, …)

Comment (YHNdnzj) [reply src/shared/battery-util.c:271]: Hmm, what if there are multiple batteries with unreadable state? We wait until the loop finishes after all 🤔

Comment (YHNdnzj) [reply src/shared/battery-util.c:271]: And I don't think we should do the logging in the loop, because we might still find charged batteries later.

Comment (poettering) [reply src/shared/battery-util.c:271]: ignore me

**Pull request merged**